src/collections/maple_tree/mod.rs

```diff
@@ -104,66 +131,307 @@ impl<K, V> SparseLeaf<K, V>
+            for i in currentIndex+1..8{
+                
+                if self.key[i] < self.key[smallestKeyIndex]{
+                    smallestKeyIndex = i;
+                }
+            }
+            
+            // if the currentIndex already holds the smallest value don't bother
+            // trying to move anything around
+            if smallestKeyIndex != currentIndex{
+
+                // these swaps are the only way I have figure out how to do this, is there an
+                // easier way? 
+                mem::swap(&mut self.key[currentIndex], &mut keyHolder);
+                mem::swap(&mut self.value[currentIndex], &mut valueHolder);
 
+                mem::swap(&mut self.key[smallestKeyIndex], &mut keyHolder);
+                mem::swap(&mut self.value[smallestKeyIndex], &mut valueHolder);
+                
+                mem::swap(&mut self.key[currentIndex], &mut keyHolder);
+                mem::swap(&mut self.value[currentIndex], &mut valueHolder);
+
+            }
+        }
+    }
     // We need to sort *just before* we split if required.
 }
 
-
 #[cfg(test)]
 mod tests {
-    use super::{SparseLeaf};
+    use super::SparseLeaf;
+    use super::M;
 
     #[test]
-    fn test_sparse_leaf_basic() {
+    fn test_sparse_leaf_search() {
         let mut sl: SparseLeaf<usize, usize> = SparseLeaf::new();
-        // insert
-        sl.insert(0, 0);
-        // remove
-        sl.remove(&0);
-        // Remove non-existant
-        sl.remove(&0);
 
-        sl.insert(1, 1);
+        // test valid search
         sl.insert(2, 2);
 
-        // Insert duplicate
-        sl.insert(2, 2); // error?
+        assert!(sl.search(&2).is_some());
+
+        // test invalid search
+        assert!(sl.search(&3).is_none());
+    }
+
+    #[test]
+    fn test_sparse_leaf_update() {
+        let mut sl: SparseLeaf<usize, usize> = SparseLeaf::new();
+
+        // Insert K:V pair
+        sl.insert(2, 2);
 
         // update inplace.
         sl.update(2, 3);
 
-        // split?
-        // compact/sort
-        // verify
-        // clone
-        // eq?
-        // pass
+        // check that the value was correctly changed
+        assert!(sl.search(&2) == Some(&3));
     }
 
-}
+    #[test]
+    fn test_sparse_leaf_insert() {
+        let mut sl: SparseLeaf<usize, usize> = SparseLeaf::new();
 
+        // insert
+        sl.insert(2, 2);
+
+        // test valid search
+        assert!(sl.search(&2) == Some(&2));
 
+        // test invalid search
+        assert!(sl.search(&1).is_none());
+
+        // test insert after node is already full
+
+        for i in 1..8 {
+            sl.insert(i, i);
+        }
+
+        assert!(sl.insert(8, 8).is_none())
+    }
+
+    #[test]
+    fn test_sparse_leaf_remove() {
+        let mut sl: SparseLeaf<usize, usize> = SparseLeaf::new();
+
+        // check removing a non-existent value fails
+        assert!(sl.remove(&0).is_none());
+
+        // check removing a value that exists
```

Comment:
Try filling the node and then remove the values out of order. IE insert 1 -> 8 then remove them like 3, 7, 8, 1, 4 ...